lib/hsa/hsaRfiUpdateBadInputs.cpp

```diff
@@ -87,64 +111,21 @@ hsa_signal_t hsaRfiUpdateBadInputs::execute(int gpu_frame_id, hsa_signal_t prece
 
 void hsaRfiUpdateBadInputs::finalize_frame(int frame_id) {
     std::lock_guard<std::mutex> lock(update_mutex);
-    if (update_bad_inputs && frames_to_update_finalize > 0) {
-        frames_to_update_finalize--;
+    
+    // Only mark input empty if filling frame and
+    // no more frames to finalize.
+    if (frame_to_fill_finalize > 0 && frame_copy_active.at(frame_id)) {
+        frame_copy_active.at(frame_id) = false;
         hsaCommand::finalize_frame(frame_id);
-        // Check if we've finished loading the new bad input mask in all frames.
-        if (frames_to_update_finalize == 0) {
-            update_bad_inputs = false;
+        DEBUG("finalize_frame for gpu_frame_id={:d} using _in_buf_finalize_id={:d}", frame_id,
+              _in_buf_finalize_id);
+        frame_to_fill_finalize--;
+        if (frame_to_fill_finalize == 0) {
+            mark_frame_empty(_in_buf, unique_name.c_str(), _in_buf_finalize_id);
```

Comment:
Since you are doing a `memcpy()` in the precondition part you can release the frame in the precondition. No need to wait for here.  That would also simply the logic.